### PR TITLE
Reshare confirmation

### PR DESCRIPTION
**Pull request by JesseAbram**
Related #941 

- [x] Allow nodes to call in to confirm
- [x]  On confirmation have chain update new signers

This PR also stores the new key and deletes the old one until the new flow is finished #968

### CODE REVIEW

File: crates/threshold-signature-server/src/validator/tests.rs [hunk 1]
```diff
@@ -77,6 +76,9 @@ async fn test_reshare() {
     for response_result in response_results {
         assert_eq!(response_result.unwrap().text().await.unwrap(), "");
     }
+    let key_share_after = unsafe_get(&client, hex::encode(NETWORK_PARENT_KEY), 3001).await;
+    assert_ne!(key_share_before, key_share_after);
```
Comment (ameba23) [inline crates/threshold-signature-server/src/validator/tests.rs:80]: I guess we could also check that bob and charlie's keyshares got updated.  Not sure if thats needed.

**Review by ameba23 — APPROVED** [review]
:100: All looks good.

The thing i don't see, which i guess just isn't in the scope of this PR, is the bit where the one member of the old signing committee who isn't in the new signing committee deletes their keyshare.

File: pallets/staking/src/lib.rs [hunk 2]
```diff
@@ -403,6 +426,43 @@ pub mod pallet {
...
+            let current_signer_length = signers_info.next_signers.len();
+            if signers_info.confirmations.len() == (current_signer_length - 1) {
+                Signers::<T>::put(signers_info.next_signers.clone());
+                Self::deposit_event(Event::SignersRotation(signers_info.next_signers));
+                Ok(Pays::No.into())
+            } else {
+                signers_info.confirmations.push(validator_stash.clone());
+                NextSigners::<T>::put(signers_info);
+                Self::deposit_event(Event::SignerConfirmed(validator_stash));
+                Ok(Pays::No.into())
```
[the remark this answers is not in the document]
Comment (JesseAbram) [reply pallets/staking/src/lib.rs:463]: free may not be the right term, refunded, as in they pay up front then if they don't hit any errors they get their fees back

File: pallets/staking/src/lib.rs [hunk 3]
```diff
@@ -222,6 +231,13 @@ pub mod pallet {
                 proactive_refresh_keys: self.proactive_refresh_data.1.clone(),
             };
             ProactiveRefresh::<T>::put(refresh_info);
+
+            if self.mock_signer_rotate {
+                NextSigners::<T>::put(NextSignerInfo {
+                    next_signers: self.inital_signers.clone(),
```
[the remark this answers is not in the document]
Comment (JesseAbram) [reply pallets/staking/src/lib.rs:237]: pretty much if next signers do not exist then the chain isn't in a rotating state and when the nodes call confirm done they will get rejected

File: pallets/staking/src/lib.rs [hunk 4]
```diff
@@ -446,7 +506,10 @@ pub mod pallet {
             // removes first signer and pushes new signer to back
             current_signers.remove(0);
             current_signers.push(next_signer_up.clone());
-            NextSigners::<T>::put(current_signers);
+            NextSigners::<T>::put(NextSignerInfo {
+                next_signers: current_signers,
+                confirmations: vec![],
+            });
```
[the remark this answers is not in the document]
Comment (JesseAbram) [reply pallets/staking/src/lib.rs:512]: yes we would need an extra node added

**Comment by JesseAbram** [comment]
> 💯 All looks good.
> 
> The thing i don't see, which i guess just isn't in the scope of this PR, is the bit where the one member of the old signing committee who isn't in the new signing committee deletes their keyshare.

good point will add it to #941

File: crates/threshold-signature-server/src/validator/tests.rs (hunk 1 above)
Comment (JesseAbram) [reply crates/threshold-signature-server/src/validator/tests.rs]: nvm it was a good idea, I just did it

**Pull request merged**